src/cm_editor.ml

```diff
@@ -0,0 +1,100 @@
+open Import
+open Interop
+
+module Cm_document : sig
+  include module type of CustomDocument
+
+  val content : t -> (string, string) result Promise.t
+
+  val create : uri:Uri.t -> t
+end = struct
+  include CustomDocument
+
+  let content t =
+    let uri = uri t in
+    let file_path = Uri.fsPath uri in
+    let bin = Path.of_string "ocamlobjinfo" in
+    Cmd.(output (Spawn { bin; args = [ file_path ] }))
+
+  let create ~(uri : Uri.t) =
+    let module EventEmitter = EventEmitter.Make (Js.Unit) in
+    let onDidChange_event_emitter = EventEmitter.make () in
+    let onDidChange_event = EventEmitter.event onDidChange_event_emitter in
+    let onDidDispose_event_emitter = EventEmitter.make () in
+    let onDidDispose_event = EventEmitter.event onDidDispose_event_emitter in
+    let watcher =
+      Workspace.createFileSystemWatcher
+        (`String (Uri.fsPath uri))
+        ~ignoreCreateEvents:true ~ignoreDeleteEvents:true ()
+    in
+    let disposable =
+      FileSystemWatcher.onDidChange watcher ~callback:(fun _ ->
+          EventEmitter.fire onDidChange_event_emitter ())
+    in
+    create ~uri ~onDidChange:onDidChange_event ~onDidDispose:onDidDispose_event
+      ~dispose:(fun () ->
+        EventEmitter.fire onDidDispose_event_emitter ();
+        Disposable.dispose disposable)
+      ()
+end
+
+let resolveCustomEditor _instance ~document ~webviewPanel ~token:_ :
+    unit Promise.t =
+  let open Promise.Syntax in
+  let update_content document =
+    let+ content = Cm_document.content document in
+    match content with
+    | Ok file_content ->
+      let webview = WebviewPanel.webview webviewPanel in
+      WebView.set_html webview (Printf.sprintf "<pre>%s</pre>" file_content)
+    | Error e ->
+      let uri = Cm_document.uri document in
+      let () = log "Error while trying to read content from file: %s" e in
+      let path = Uri.path uri in
+      let (_ : 'a option Promise.t) =
+        Window.showErrorMessage
+          ~message:
+            (Printf.sprintf "Error while trying to read content from %s."
+               (Node.Path.basename path))
+          ()
+      in
+      ()
+  in
+  let disposable =
+    Cm_document.onDidChange document
+      ~listener:(fun () ->
+        let (_ : unit Promise.t) = update_content document in
+        ())
+      ()
+  in
+  let (_ : Disposable.t) =
+    WebviewPanel.onDidDispose webviewPanel
+      ~listener:(fun () -> Disposable.dispose disposable)
+      ()
+  in
+  update_content document
+
+let openCustomDocument _instance ~(uri : Uri.t) ~openContext:_ ~token:_ :
+    Cm_document.t Promise.t =
+  let document = Cm_document.create ~uri in
+  Promise.resolve document
+
+let register (extension : ExtensionContext.t) (instance : Extension_instance.t)
+    =
+  let module CustomReadonlyEditorProvider =
+    CustomReadonlyEditorProvider.Make (Cm_document) in
```

Comment:
According to the TS types, `CustomReadonlyEditorProvider` works only with [objects that implement the `CustomDocument` interface](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/vscode/index.d.ts#L8188).

The functor here accepts any modules, not only those that implement a specific interface. Not sure if there's a way to enforce that in OCaml?

Reply:
Would this work?

```ocaml
module Make (T : module type of CustomDocument) = ...
```

Reply:
When I do that I've another error elsewhere that says:

```bash
File "src-bindings/vscode/vscode.ml", line 3174, characters 41-42:
3174 |       CustomReadonlyEditorProvider.Make (T) in
                                                ^
Error: Signature mismatch:
       ...
       The value `create' is required but not provided
       File "src-bindings/vscode/vscode.ml", line 2353, characters 4-80:
         Expected declaration
       The value `dispose' is required but not provided
       File "src-bindings/vscode/vscode.ml", line 2351, characters 4-39:
         Expected declaration
       The value `uri' is required but not provided
       File "src-bindings/vscode/vscode.ml", line 2349, characters 4-35:
         Expected declaration
```

I think that makes sense given how `registerCustomReadonlyEditorProvider` is defined:

```ocaml
let registerCustomReadonlyEditorProvider (type a)
      (module T : Js.T with type t = a) ~(viewType : string)
      ~(provider : a CustomReadonlyEditorProvider.t)
      ?(options : RegisterCustomEditorProviderOptions.t or_undefined) () :
      Disposable.t =
    let module CustomReadonlyEditorProvider =
      CustomReadonlyEditorProvider.Make (T) in
    let provider = [%js.of: CustomReadonlyEditorProvider.t] provider in
    registerCustomReadonlyEditorProvider ~viewType ~provider ?options ()
```

I tried to create a `module type T = sig ...` inside of the `CustomDocument` module so that I could change `(module T : Js.T with type t = a)` by `(module T : CustomDocument.T with type t = a)` but then I've another error that I have no clue how to fix 😅

Reply:
I see. I think adding the `T` module type is the correct solution. What is the error that you get if you include it?

Reply:
There was two errors. I fixed one by adding the `dispose` argument to the `Cm_document.create` function to match the `CustomDocument.create` function (which I think makes sense).

The second error is:

```ocaml
File "src-bindings/vscode/vscode.ml", line 3174, characters 41-42:
3174 |       CustomReadonlyEditorProvider.Make (T) in
                                                ^
Error: Signature mismatch:
       ...
       The module type `T' is required but not provided
       File "src-bindings/vscode/vscode.ml", lines 2331-2343, characters 2-5:
         Expected declaration
make: *** [all] Error 1    
```

I don't know how to fix that one... I pushed a commit https://github.com/ocamllabs/vscode-ocaml-platform/pull/798/commits/2c60512e0d9f72fd5886996ed49517978e865aeb with the changes if you want to have a look.

Reply:
You need to change `module type of CustomDocument` to `CustomDocument.T` here:
https://github.com/ocamllabs/vscode-ocaml-platform/blob/2c60512e0d9f72fd5886996ed49517978e865aeb/src-bindings/vscode/vscode.ml#L2905

Reply:
Oops 🤦‍♂️ I also had to apply the following change and now it works! 🥳 Thanks for your help! 🤗❤️

```diff
@@ -2373,7 +2373,7 @@ module Window : sig
     -> Disposable.t
 
   val registerCustomReadonlyEditorProvider :
-       'a Js.t
+       (module CustomDocument.T with type t = 'a)
     -> viewType:string
     -> provider:'a CustomReadonlyEditorProvider.t
     -> ?options:RegisterCustomEditorProviderOptions.t
```